    We study the classic problem of inferring ancestral genomes under a given phylogeny and from a given set of extant genomes, also known as the \emph{small parsimony problem} (SPP).
    The evolutionary model considered at hand encompasses large scale rearrangements acting on so-called \emph{gene order} sequences and includes segmental gain and loss. Each genome may encompass one or more linear or circular chromosomes, and genomic markers may appear in several copies, without restriction on their genomic location or orientation within the sequence.
    Therefore, our method presumes ancestral copy number estimates, specified either as fixed count, or as range. 
    In case of the latter, our method chooses the copy number within the range that minimizes the total number of large scale rearrangements over the entire phylogeny. 
    
Even under simple evolutionary models, such as the classic character-state model, the SPP is computationally intractable.
However, we give a highly optimized ILP that is able to compute the SPP under our model for sufficiently small phylogenies and gene families. We benchmark our method on simulated phylogenies and discuss its performance in reconstructing gene orders under our broad evolutionary model. 

