% This is samplepaper.tex, a sample chapter demonstrating the
% LLNCS macro package for Springer Computer Science proceedings;
% Version 2.21 of 2022/01/12
%
\documentclass[runningheads]{llncs}
%
\input{preamble}
\usepackage[T1]{fontenc}
% T1 fonts will be used to generate the final print and online PDFs,
% so please use T1 fonts in your manuscript whenever possible.
% Other font encondings may result in incorrect characters.
%
\usepackage{graphicx}
% Used for displaying a sample figure. If possible, figure files should
% be included in EPS format.
%
% If you use the hyperref package, please uncomment the following two lines
% to display URLs in blue roman font according to Springer's eBook style:

%\renewcommand\UrlFont{\color{blue}\rmfamily}
%

\begin{document}
%
\title{Reconstructing Rearrangement Phylogenies of Natural Genomes}
%
%\titlerunning{Abbreviated paper title}
% If the paper title is too long for the running head, you can set
% an abbreviated paper title here
%
\author{Leonard Bohnen\"amper\inst{1}\orcidID{0000-0003-4508-0078} \and
    Jens Stoye \inst{1}\orcidID{0000-0002-4656-7155} \and
    Daniel D\"orr \inst{2}\orcidID{0000-0002-3720-6227}
}
%
\authorrunning{L. Bohnenk\"amper et al.}
% First names are abbreviated in the running head.
% If there are more than two authors, 'et al.' is used.
%
\institute{Faculty of Technology and Center for Biotechnology (CeBiTec), Bielefeld University, Germany\\ 
    \email{lbohnenkaemper@techfak.uni-bielefeld.de}, \email{jens.stoye@uni-bielefeld.de}\\
\and
Institute for Medical Biometry and Bioinformatics, Medical Faculty, and Center for Digital Medicine, Heinrich Heine University, Germany\\
\email{daniel.doerr@hhu.de}}

%
\maketitle              % typeset the header of the contribution
%
\begin{abstract}

\input{abstract}


\keywords{genome rearrangement \and ancestral reconstruction \and small parsimony \and integer linear programming \and double-cut-and-join}
\end{abstract}




\section{Introduction}
\section{Background}

\subsection{Preliminaries}
A \emph{(genomic) marker} $g := (g^\et, g^\eh)$ is a universally unique entity consisting of \emph{marker extremities} tail of $g$, denoted by $g^\et$, and head of $g$, denoted by $g^\eh$. 
A \emph{telomere} $t^\circ$ is a universally unique entity encompassing a single telomeric extremity denoted by ``$\circ$''. 
A genome is a triple $(M \cup \pseudocaps, A, \family)$, with $M$ being its marker set and $\pseudocaps$ its set of telomeres, $A \subset (M \cup \pseudocaps)^2$ the set of telomeres, and $\family: M \to \mathbb N$ a function indicating the family of each marker. 
The set of adjacencies $A$ has the properties that \emph{(i)} for each extremity $\varepsilon \in \bigcup M \cup \pseudocaps$, there exists exactly one adjacency $a \in A$ that contains $\varepsilon$. 
The adjacency set decomposes the set of markers and telomeres into linear or circular components, that we further call \emph{chromosomes}. 
Moreover, \emph{(ii)} each linear chromosome, formed by ordering markers and telomeres based on their adjacencies in $A$, starts and ends with a telomere. 

A \emph{(genome) scaffold} is a quintuple $(M \cup \pseudocaps, A, \family, \multi, \w)$ with 
\begin{itemize}
    \item $M$, $\pseudocaps$, and $\family$ as defined in a genome, 
    \item the set of adjacencies $A$ satisfying that for each $g^\et \in \bigcup A$, there exists also extremity $g^\eh \in \bigcup A$ and vice versa, and each telomeric extremity is used only once, i.e., $\forall \{X, X'\} \subseteq A$, $X \cap X' \cap \pseudocaps = \emptyset$,
    \item copy number function $\multi(i) = [a, b]$ reporting for each family $i$ the permitted range $0 \leq a \leq b$ of its copy number, such that each family, defined as $F_i := \{g \in M \mid f(g) = i\}$, satisfies $|F_i| = b$, and
    \item adjacency weight function $\w: A \to \mathbb R$. % [-1, 1]$.
\end{itemize}

Observe that a genome can give rise to a scaffold, with copy number $\multi$ and weight $\w$ functions arbitrarily defined, but the reverse does not hold true in general.
% not sure we need this:
% However, in the following, we regard genomes as scaffolds with tight copy numbers, that is, the copy number range of each family $i$ corresponds to a point that coincides with the number of its associated markers.
A genome $S' = (M', A', \family')$ is \emph{derived} from a scaffold $S = (M, A, \family, \multi, \w)$, or simply ``$S$-derived'', if \emph{(i)} $M' \subseteq M$, \emph{(ii)} $A' \subseteq A$, \emph{(iii)} 
there exist no two markers $g, g' \in M'$ with $\family'(g) = \family'(g')$ and $\family(g) \neq \family(g')$, and \emph{(iv)} the original copy number constraints are satisfied, i.e., for each family $i$ and the set of markers associated with $i$ in scaffold $S$, i.e., $F_i = \{ g \in M' \mid \family(g) = i\}$, holds true that $|F_i| \in \multi(i)$.
We call a scaffold $S$ \emph{linearizable} if there exists an $S$-derived genome. 
In fact, many scaffolds are not linearizable. 

A \emph{phylogeny} $\tree$ is a connected graph with nodes representing \emph{operational taxonomic units} (OTUs). 
Nodes with degree 1 are termed \emph{tips} of the phylogeny.

General notation
\begin{itemize}
    \item 
    \item Weighted multigraph $G = (\vertices, \edges, \w)$ with edge weight function $\w: \edges \to \mathbb R$
\end{itemize}

Paper-specific notation:

\begin{itemize}
    \item Genomic marker 
    \item Furthermore, we use a function $\extf : \extremities \to \{\et, \eh, \circ\}$ to map extremities to their corresponding kind (tail, head or telomere). 
    \item 
We model family assignments of marker extremities as a function $\family: \extremities \to \mathbb N$ for which holds true that for any marker $g = \{g^\et, g^\eh\}$, $\family(g^\et) = \family(g^\eh)$.  
Function 
$\minmulti_\X : \mathbb N \to \mathbb N$ reports the minimum multiplicity of a gene family in a given genome $\X$, while $\maxmulti_\X: \mathbb N \to \mathbb N$ reports its maximum multiplicity. 
    \item Multirelational diagram $\mrd$ and \emph{capping-free multi-relational diagram} (CFMRD) $\pcmrd$
\end{itemize}

\begin{itemize}
    \item SPP algoirthms such as 
\end{itemize}

\begin{problem}[Weighted CN-constrained degenerate DCJ indel distance]\label{prb:wdeg_dcj}
    Given a weighting scheme $\w : \extremities \times \extremities \to \mathbb R$, some $\alpha \in [0, 1]$, two linearizable degenerate genomes $\A, \B$ with copy number constraints \hl{XX}\todoD{add data structure} and family assignment $\family$, find $\A$-derived genome $\A'$, $\B$-derived genome $\B'$, and $\family$-derived $\{\A',\B'\}$-resolved family assignment $\family'$ that minimize the linear combination
    $$
    (1-\alpha) \cdot \sum_{X \in \A' \cup \B'} -\w(X) + \alpha \cdot \dist_\DCJid(A', B')\,.
    $$
\end{problem}

\begin{problem}[SPP-DCJ]\label{prb:spp_dcj}
    Given a phylogeny $\tree$ and a set of linearizable degenerate genomes $\A_1, \ldots, \A_k$ corresponding to the node set $V(\tree) = \{\A_1, \ldots, \A_k\}$, find genomes $\A'_1 \subseteq \A_1, \ldots, \A'_k \allowbreak \subseteq \A_k$ that minimize the sum of weighted degenerate DCJ indel distances along the edges of $\tree$. 
\end{problem}

\section{A new Method}
\subsection{Pre-selecting Adjacencies}
\todoD{Describe Process}\subsection{A New ILP Formulation}



The algorithm described in the following operates on two levels: 
on the local level, a genome is derived from each GAG, constituting a set of linear or circular chromosomes. 
On the global level, genomes are connected to each other along the branches of the phylogeny. Each branch gives rise to a pairwise comparison by means of the CFMRD. 
In doing so, the selection of adjacencies of a derived genome is propagated from across CFMRDs, thus ensuring global consistency. 

\paragraph{Local level.} 
Constraint \ref{c:cn} implements the user-provided copy number requirement. 
In doing so, we define the set of tail extremities of markers from family $i$ in GAG $\ggraph(\X)$ as $F_i = \{ v \mid v \in V: \family(v) = i \text{ and } \extf(v) = \et\}$ and iterate over the associated $\ilpvar{g}$ variables of its members to bound the number of active markers within the given range. 
The choice of counting tail extremities is arbitrary, however, the subsequent constraint (\ref{c:cn_consistent}) ensures that the assignment of $\ilpvar{g}$ variables for both extremities of a markers is consistent. 
At last, Constraint \ref{c:genome} makes sure that a valid genome is derived from the GAG when solving the ILP, by enforcing that each extremity node of $\ggraph(\X)$ is incident to at most one adjacency edge. 

\paragraph{Global level.}

When comparing pairs of genomes within a CFMRD, we make use of a capping-free formulation for the computation of the pairwise DCJ indel distance~\cite{BOH-2023}.


\todoL{Insert a description of the formulation here}

\include{ilp}


\section{Evaluation}
\section{Discussion}



\subsubsection{Acknowledgements} Please place your acknowledgments at
the end of the paper, preceded by an unnumbered run-in heading (i.e.
3rd-level heading).

%
% ---- Bibliography ----
%
% BibTeX users should specify bibliography style 'splncs04'.
% References will then be sorted and formatted in the correct style.
%
\bibliographystyle{splncs04}
\bibliography{refs}


\end{document}
