% This is samplepaper.tex, a sample chapter demonstrating the
% LLNCS macro package for Springer Computer Science proceedings;
% Version 2.21 of 2022/01/12
%
\documentclass[runningheads]{llncs}
%
\input{preamble}
\usepackage[T1]{fontenc}
% T1 fonts will be used to generate the final print and online PDFs,
% so please use T1 fonts in your manuscript whenever possible.
% Other font encondings may result in incorrect characters.
%
\usepackage{graphicx}
% Used for displaying a sample figure. If possible, figure files should
% be included in EPS format.
%
% If you use the hyperref package, please uncomment the following two lines
% to display URLs in blue roman font according to Springer's eBook style:

%\renewcommand\UrlFont{\color{blue}\rmfamily}
%

\begin{document}
%
\title{Reconstructing Rearrangement Phylogenies of Natural Genomes}
%
%\titlerunning{Abbreviated paper title}
% If the paper title is too long for the running head, you can set
% an abbreviated paper title here
%
\author{Leonard Bohnen\"amper\inst{1}\orcidID{0000-0003-4508-0078} \and
    Jens Stoye \inst{1}\orcidID{0000-0002-4656-7155} \and
    Daniel D\"orr \inst{2}\orcidID{0000-0002-3720-6227}
}
%
\authorrunning{L. Bohnenk\"amper et al.}
% First names are abbreviated in the running head.
% If there are more than two authors, 'et al.' is used.
%
\institute{Faculty of Technology and Center for Biotechnology (CeBiTec), Bielefeld University, Germany\\ 
    \email{lbohnenkaemper@techfak.uni-bielefeld.de}, \email{jens.stoye@uni-bielefeld.de}\\
\and
Institute for Medical Biometry and Bioinformatics, Medical Faculty, and Center for Digital Medicine, Heinrich Heine University, Germany\\
\email{daniel.doerr@hhu.de}}

%
\maketitle              % typeset the header of the contribution
%
\begin{abstract}

\input{abstract}


\keywords{genome rearrangement \and ancestral reconstruction \and small parsimony \and integer linear programming \and double-cut-and-join}
\end{abstract}




%\section{Introduction}
\section{Background}

\begin{itemize}
    \item SPP algoirthms such as 
\end{itemize}
\section{A new Method}
\subsection{Preliminaries}
General notation
\begin{itemize}
    \item Weighted multigraph $G = (\vertices, \edges, \w)$ with edge weight function $\w: \edges \to \mathbb R$
    \item A \emph{phylogeny} $\tree$ is a connected graph. Nodes of $\tree$ with degree 1 are termed \emph{leaves}.
\end{itemize}

Paper-specific notation:

\begin{itemize}
    \item Genomic marker 
    \item Furthermore, we use a function $\extf : \extremities \to \{\et, \eh, \circ\}$ to map extremities to their corresponding kind (tail, head or telomere). 
    \item Genome $\X$
    \item Genome adjacency graph $\ggraph(\X) = (\vertices, \edges, \family)$ \todoD{find better name?}
    \item 
We model family assignments of marker extremities as a function $\family: \extremities \to \mathbb N$ for which holds true that for any marker $g = \{g^\et, g^\eh\}$, $\family(g^\et) = \family(g^\eh)$.  
Function 
$\minmulti_\X : \mathbb N \to \mathbb N$ reports the minimum multiplicity of a gene family in a given genome $\X$, while $\maxmulti_\X: \mathbb N \to \mathbb N$ reports its maximum multiplicity. 
    \item Multirelational diagram $\mrd$ and \emph{capping-free multi-relational diagram} (CFMRD) $\pcmrd$
\end{itemize}
\subsection{Pre-selecting Adjacencies}
\todoD{Describe Process}
\subsection{A New ILP Formulation}
The algorithm described in the following operates on two levels: 
on the local level, a genome is derived from each GAG, constituting a set of linear or circular chromosomes. 
On the global level, genomes are connected to each other along the branches of the phylogeny. Each branch gives rise to a pairwise comparison by means of the CFMRD. 
In doing so, the selection of adjacencies of a derived genome is propagated from across CFMRDs, thus ensuring global consistency. 

\paragraph{Local level.} 
Constraint \ref{c:cn} implements the user-provided copy number requirement. 
In doing so, we define the set of tail extremities of markers from family $i$ in GAG $\ggraph(\X)$ as $F_i = \{ v \mid v \in V: \family(v) = i \text{ and } \extf(v) = \et\}$ and iterate over the associated $\ilpvar{g}$ variables of its members to bound the number of active markers within the given range. 
The choice of counting tail extremities is arbitrary, however, the subsequent constraint (\ref{c:cn_consistent}) ensures that the assignment of $\ilpvar{g}$ variables for both extremities of a markers is consistent. 
At last, Constraint \ref{c:genome} makes sure that a valid genome is derived from the GAG when solving the ILP, by enforcing that each extremity node of $\ggraph(\X)$ is incident to at most one adjacency edge. 

\paragraph{Global level.}

When comparing pairs of genomes within a CFMRD, we make use of a capping-free formulation for the computation of the pairwise DCJ indel distance~\cite{BOH-2023}.


\todoL{Insert a description of the formulation here}

\include{ilp}


\section{Evaluation}
\section{Discussion}



\subsubsection{Acknowledgements} Please place your acknowledgments at
the end of the paper, preceded by an unnumbered run-in heading (i.e.
3rd-level heading).

%
% ---- Bibliography ----
%
% BibTeX users should specify bibliography style 'splncs04'.
% References will then be sorted and formatted in the correct style.
%
\bibliographystyle{splncs04}
\bibliography{refs}


\end{document}
